consensus/src/dag/adapter.rs

```diff
@@ -87,3 +98,164 @@ impl Notifier for BufferManagerAdapter {
         })?)
     }
 }
+
+pub struct StorageAdapter {
+    epoch: u64,
+    epoch_to_validators: HashMap<u64, Vec<Author>>,
+    consensus_db: Arc<ConsensusDB>,
+    aptos_db: Arc<dyn DbReader>,
+}
+
+impl StorageAdapter {
+    pub fn new(
+        epoch: u64,
+        epoch_to_validators: HashMap<u64, Vec<Author>>,
+        consensus_db: Arc<ConsensusDB>,
+        aptos_db: Arc<dyn DbReader>,
+    ) -> Self {
+        Self {
+            epoch,
+            epoch_to_validators,
+            consensus_db,
+            aptos_db,
+        }
+    }
+
+    pub fn bitvec_to_validators(
+        validators: &[Author],
+        bitvec: &BitVec,
+    ) -> anyhow::Result<Vec<Author>> {
+        if BitVec::required_buckets(validators.len() as u16) != bitvec.num_buckets() {
+            bail!(
+                "bitvec bucket {} does not match validators len {}",
+                bitvec.num_buckets(),
+                validators.len()
+            );
+        }
+
+        Ok(validators
+            .iter()
+            .enumerate()
+            .filter_map(|(index, validator)| {
+                if bitvec.is_set(index as u16) {
+                    Some(*validator)
+                } else {
+                    None
+                }
+            })
+            .collect())
+    }
+
+    pub fn indices_to_validators(
+        validators: &[Author],
+        indices: &[u64],
+    ) -> anyhow::Result<Vec<Author>> {
+        indices
+            .iter()
+            .map(|index| {
+                usize::try_from(*index)
+                    .map_err(|_err| anyhow!("index {} out of bounds", index))
```

Comment:
this should be an assertion as well

Reply:
yeah those code was from leader reputation where we don't need to panic